lib/analyzer/image-extractor.ts

```diff
@@ -60,51 +65,64 @@ async function extractLayerKeyFiles(
       // all layer level entries read
       resolve(result);
     });
-    stream.pipe(layerExtract);
+    layerTarStream.pipe(layerExtract);
   });
 }
 
 /**
  * Extract key files textual content and MD5 sum from the specified TAR file.
  * @param imageTarPath path to image file saved in tar format
- * @param txtPatterns list of plain text key files paths patterns to extract and return as strings
- * @param md5Patterns list of binary key files paths patterns to extract and return their MD5 sum
- * @returns manifest file and key files by inner layers
+ * @param lookups array of pattern, callback pairs
+ * @returns extracted files products
  */
-async function extractImageKeyFiles(
+async function extractFromTar(
   imageTarPath: string,
-  txtPatterns: string[],
-  md5Patterns: string[] = [],
-): Promise<ExtractedImage> {
+  lookups: LookupEntry[],
+): Promise<ExtractedFiles> {
   return new Promise((resolve) => {
     const imageExtract: Extract = extract();
-    let manifest: string;
-    const layers: Layers = {};
+    const layers: ExtractedFilesByLayers = {};
+    let layersNames: string[];
 
     imageExtract.on("entry", (header, stream, next) => {
       if (header.type === "file") {
         if (basename(header.name) === "layer.tar") {
-          extractLayerKeyFiles(stream, txtPatterns, md5Patterns).then(
-            (extractedKeyFiles) => {
-              layers[header.name] = extractedKeyFiles;
-            },
-          );
+          extractFromLayer(stream, lookups).then((extractedKeyFiles) => {
+            layers[header.name] = extractedKeyFiles;
+          });
         } else if (header.name === "manifest.json") {
           streamToString(stream).then((manifestFile) => {
-            manifest = manifestFile;
+            const manifest = JSON.parse(manifestFile);
+            layersNames = manifest[0].Layers;
           });
         }
       }
       stream.resume(); // auto drain the stream
       next(); // ready for next entry
     });
+
     imageExtract.on("finish", () => {
-      // all image level entries read
-      resolve({
-        manifest,
-        layers,
-      });
+      const result: ExtractedFiles = {};
+
+      if (layers) {
+        // reverse layer order from last to first
+        for (const layerName of layersNames.reverse()) {
```

Comment:
please change the `if`s statements, so you will have less indentations.
Instead of:
`if (x) { if (y) {return z;}` you can do `if (!x) {return} if (y) {return z}`

Reply:
Done explicitly by adding `if (!layers) { return {}; }` and implicitly after the split into separate function made 2 of the loops redundant.